### PR TITLE
Allow users to update existing autopilot rules

### PR DESCRIPTION
**Pull request by trierra**
Signed-off-by: trierra <oksana@portworx.com>

### CODE REVIEW

File: pkg/rule/engine/v1/engine.go [hunk 1]
```diff
@@ -280,3 +308,11 @@ func init() {
 		logrus.Panicf("failed to register engine %s due to : %v", Name, err)
 	}
 }
+
+func equal(existingRuleSpec *v1alpha1.AutopilotRuleSpec, autopilotRule *v1alpha1.AutopilotRuleSpec) bool {
```
Comment (adityadani) [inline pkg/rule/engine/v1/engine.go]: nit: as this is a global function, can we rename this to areSpecsSame or areSpecsEqual ?

File: pkg/rule/rule.go [hunk 2]
```diff
@@ -179,6 +181,19 @@ func (r *ruleObjectInstance) Check() ([]*v1alpha1.LabelSelectorRequirement, erro
 	return matchedConditions, nil
 }
 
+// UpdateParentsForRule updates parents rule in rule object. Required in order to support RuleUpdate feature
+func (r *ruleObjectInstance) UpdateParentsForRule(newParent *v1alpha1.AutopilotRule) error {
+	parentUUID := GetUUIDFromRule(newParent)
+	// checking if new parentUUID matches the existing one to avoid accidental setting wrong parents
+	if GetUUIDFromRule(r.parentRule) != parentUUID {
+		return fmt.Errorf("can't update parent for rule %v. New ParentUUID '%s' doesn't match the existing ParentUUID '%s'",
```
Comment (adityadani) [inline pkg/rule/rule.go]: nit: cannot update parent for rule...

File: pkg/rule/rule.go [hunk 3]
```diff
@@ -179,6 +181,19 @@ func (r *ruleObjectInstance) Check() ([]*v1alpha1.LabelSelectorRequirement, erro
...
+			r.UUID(), parentUUID, GetUUIDFromRule(r.parentRule))
+	}
+	r.parentRule = newParent
```
Comment (adityadani) [inline pkg/rule/rule.go]: we need a lock while updating the parent. the same rule object could be getting dequeued and processed at the same time when the parent rule is getting updated.

Comment (trierra) [reply pkg/rule/rule.go]: right, thanks

**Pull request merged**